File: leetcode/src/binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        if (root is None):
            return []
    
        frontier = [root]
        result = []
        while (len(frontier) != 0):
            nextFrontier = []
            result.append([])
            for node in frontier:
                result[-1].append(node.val)
                if (node.left != None):
                    nextFrontier.append(node.left)
                if (node.right != None):
                    nextFrontier.append(node.right)
            
            frontier = nextFrontier

        return result        

File: leetcode/src/binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (root == nullptr) {
            return {};
        }
        vector<vector<int>> result;

        vector<TreeNode*> frontier {root};

        while (frontier.size() != 0) {
            vector<TreeNode*> newFrontier;
            result.push_back({});
            for (TreeNode* node : frontier) {
                result[result.size()-1].push_back(node->val);

                if (node->left != nullptr) {
                    newFrontier.push_back(node -> left);
                }
                if (node->right != nullptr) {
                    newFrontier.push_back(node -> right);
                }
                
            }

            frontier = std::move(newFrontier);
        }

        return result;
    }
};